### PR TITLE
prepublish-on-install changes

### PR DESCRIPTION
**Pull request by othiym23**
See #10074 for details.

Implemented:
- [x] `prepare` lifecycle event
- [x] `prepublishOnly` lifecycle event
- [x] deprecation warning for regular `prepublish`
- [x] documentation

**r**: @iarna, @zkat

### CODE REVIEW

File: doc/misc/npm-scripts.md [hunk 1]
```diff
@@ -6,9 +6,14 @@ npm-scripts(7) -- How npm handles the "scripts" field
 npm supports the "scripts" property of the package.json script, for the
 following scripts:
 
+* prepare:
+  Run both before the package is published, and on local `npm
+  install` without any arguments. See below.)
```
Comment (othiym23) [inline doc/misc/npm-scripts.md]: good catch! thanks!

File: doc/misc/npm-scripts.md [hunk 2]
```diff
@@ -6,9 +6,14 @@ npm-scripts(7) -- How npm handles the "scripts" field
 npm supports the "scripts" property of the package.json script, for the
 following scripts:
 
+* prepare:
+  Run both before the package is published, and on local `npm
+  install` without any arguments. (See below.)
 * prepublish:
   Run BEFORE the package is published.  (Also run on local `npm
-  install` without any arguments.)
+  install` without any arguments. See below.)
```
Comment (watilde) [inline doc/misc/npm-scripts.md]: The last part is also missing opening parenthesis

**Comment by watilde** [comment]
It's better to mention about the prepare script in the docs of the package.json as well, isn't it?
https://github.com/npm/npm/blob/master/doc/files/package.json.md#devdependencies

**Comment by watilde** [comment]
I'm wondering if we can use this new action by ourselves: https://github.com/npm/npm/blob/master/package.json#L203

File: doc/misc/npm-scripts.md (hunk 2 above)
Comment (othiym23) [inline doc/misc/npm-scripts.md]: It's on the previous line!

**Review by iarna — APPROVED** [review]
Looks good

File: test/tap/prepublish.js [hunk 3]
```diff
@@ -68,11 +68,45 @@ test('test', function (t) {
...
+      '> echo ok\\r?\\n' +
+      '\\r?\\n' +
+      'ok' +
+      '', 'ig')
+
+    t.match(c, regex)
+    t.end()
+  })
+})
 test('cleanup', function (t) {
```
Comment (othiym23) [inline test/tap/prepublish.js]: blank line above here? only the tiniest of nits to pick!

**Comment by othiym23** [comment]
@zkat's changes LGTM. :sheep::rocket::fire:

**Comment by SimenB** [comment]
I get from reading the code (I think) the `prepare` runs before `prepublishOnly` (upper case still looks weird to me), but could that be specified in the docs?

File: doc/files/package.json.md [hunk 4]
```diff
@@ -517,7 +517,7 @@ from the root of a package, and can be managed like any other npm
 configuration param.  See `npm-config(7)` for more on the topic.
 
 For build steps that are not platform-specific, such as compiling
-CoffeeScript or other languages to JavaScript, use the `prepublish`
+CoffeeScript or other languages to JavaScript, use the `prepare`
```
Comment (mightyiam) [inline doc/files/package.json.md]: Why not advise here to use `prepublishOnly`?

Comment (zkat) [inline doc/files/package.json.md]: This use case is what `prepare` is intended for: preparing the package for use. `prepublishOnly` is explicitly for the act of taking the contents of a project and putting them on the registry. The distinction matters when you use `npm install` without arguments.

File: doc/misc/npm-scripts.md [hunk 5]
```diff
@@ -43,7 +48,30 @@ names will be run for those as well (e.g. `premyscript`, `myscript`,
...
-## COMMON USES
+## PREPUBLISH AND PREPARE
+
+### DEPRECATION NOTE
+
+Since `npm@1.1.71`, the npm CLI has run the `prepublish` script for both `npm
+publish` and `npm install`, because it's a convenient way to prepare a package
+for use (some common use cases are described in the section below).  It has
+also turned out to be, in practice, [very
+confusing](https://github.com/npm/npm/issues/10074).  As of `npm@4.0.0`, a new
```
Comment (zkat) [inline doc/misc/npm-scripts.md]: It is a constant struggle. I love my team, but they still double-space after their periods, in an age of kerning. 😭

**Comment by zkat** [comment]
@SimenB that could be a useful doc patch!

Also: yes. `pare` is a thing now. 😂

**Pull request merged**